File: 29Items-Projects/7Algorithms/6-Backtracking/python/cryptarithmetic.py
"""Cryptarithmetic — solve puzzles like SEND + MORE = MONEY."""


class Cryptarithmetic:
    def __init__(self, word1: str, word2: str, result: str):
        self.word1 = word1.upper()
        self.word2 = word2.upper()
        self.result = result.upper()

        seen: set[str] = set()
        self.letters: list[str] = []
        for ch in self.word1 + self.word2 + self.result:
            if ch not in seen:
                seen.add(ch)
                self.letters.append(ch)

        self.assignment: dict[str, int] = {}
        self.solutions: list[dict[str, int]] = []

    def solve_and_print(self) -> None:
        print(f"Puzzle: {self.word1} + {self.word2} = {self.result}")
        print(f"Unique letters: {self.letters}\n")

        self._backtrack(0, set())
        if not self.solutions:
            print("No solution found.")
        else:
            print(f"Found {len(self.solutions)} solution(s):\n")
            for i, sol in enumerate(self.solutions, 1):
                print(f"Solution #{i}:")
                print(f"  {self._format(self.word1, sol)}")
                print(f"+ {self._format(self.word2, sol)}")
                print(f"= {self._format(self.result, sol)}\n")

    def _backtrack(self, idx: int, used: set[int]) -> None:
        if idx == len(self.letters):
            if self._check():
                self.solutions.append(dict(self.assignment))
            return

        letter = self.letters[idx]
        is_leading = letter in (self.word1[0], self.word2[0], self.result[0])

        for digit in range(1 if is_leading else 0, 10):
            if digit not in used:
                self.assignment[letter] = digit
                used.add(digit)
                self._backtrack(idx + 1, used)
                del self.assignment[letter]
                used.discard(digit)

    def _check(self) -> bool:
        v1 = self._to_number(self.word1)
        v2 = self._to_number(self.word2)
        vr = self._to_number(self.result)
        return v1 + v2 == vr

    def _to_number(self, word: str) -> int:
        num = 0
        for ch in word:
            num = num * 10 + self.assignment[ch]
        return num

    def _format(self, word: str, sol: dict[str, int]) -> str:
        return "".join(str(sol[ch]) for ch in word)

File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/cryptarithmetic/Cryptarithmetic.java
package cryptarithmetic;

import java.util.*;

public class Cryptarithmetic {

    private final String word1;
    private final String word2;
    private final String result;
    private final List<Character> uniqueLetters;
    private final Map<Character, Integer> assignment = new HashMap<>();
    private final List<Map<Character, Integer>> solutions = new ArrayList<>();

    public Cryptarithmetic(String word1, String word2, String result) {
        this.word1 = word1.toUpperCase();
        this.word2 = word2.toUpperCase();
        this.result = result.toUpperCase();

        Set<Character> letters = new LinkedHashSet<>();
        for (char c : this.word1.toCharArray()) letters.add(c);
        for (char c : this.word2.toCharArray()) letters.add(c);
        for (char c : this.result.toCharArray()) letters.add(c);
        this.uniqueLetters = new ArrayList<>(letters);
    }

    public void solveAndPrint() {
        System.out.printf("Puzzle: %s + %s = %s%n", word1, word2, result);
        System.out.printf("Unique letters: %s%n%n", uniqueLetters);

        boolean[] usedDigits = new boolean[10];
        backtrack(0, usedDigits);

        if (solutions.isEmpty()) {
            System.out.println("No solution found.");
        } else {
            System.out.printf("Found %d solution(s):%n%n", solutions.size());
            for (int i = 0; i < solutions.size(); i++) {
                Map<Character, Integer> sol = solutions.get(i);
                System.out.printf("Solution #%d:%n", i + 1);
                System.out.printf("  %s%n", formatWord(word1, sol));
                System.out.printf("+ %s%n", formatWord(word2, sol));
                System.out.printf("= %s%n%n", formatWord(result, sol));
            }
        }
    }

    private void backtrack(int letterIndex, boolean[] usedDigits) {
        if (letterIndex == uniqueLetters.size()) {
            if (checkEquation()) {
                solutions.add(new HashMap<>(assignment));
            }
            return;
        }

        char letter = uniqueLetters.get(letterIndex);
        boolean isFirstLetter = (letter == word1.charAt(0))
                || (letter == word2.charAt(0))
                || (letter == result.charAt(0));

        int start = isFirstLetter ? 1 : 0;
        for (int digit = start; digit <= 9; digit++) {
            if (!usedDigits[digit]) {
                assignment.put(letter, digit);
                usedDigits[digit] = true;
                backtrack(letterIndex + 1, usedDigits);
                assignment.remove(letter);
                usedDigits[digit] = false;
            }
        }
    }

    private boolean checkEquation() {
        long val1 = wordToNumber(word1);
        long val2 = wordToNumber(word2);
        long res = wordToNumber(result);
        return val1 + val2 == res;
    }

    private long wordToNumber(String word) {
        long num = 0;
        for (char c : word.toCharArray()) {
            num = num * 10 + assignment.get(c);
        }
        return num;
    }

    private String formatWord(String word, Map<Character, Integer> sol) {
        StringBuilder sb = new StringBuilder();
        for (char c : word.toCharArray()) {
            sb.append(sol.get(c));
        }
        return sb.toString();
    }
}
